### PR TITLE
support the use of different C++ standard libraries

### PR DESCRIPTION
**Pull request by inabao**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -74,7 +69,7 @@ else()
 endif()
 
 if (ENABLE_LIBCXX)
-    message(STATUS "Using libc++")
+    message(STATUS "Using libc++ with std::__1::basic_string")
```
Comment (ZhitaoShen) [inline CMakeLists.txt]: message(STATUS "Using libc++ with C++11 ABI") 

This message looks better?

File: CMakeLists.txt [hunk 2]
```diff
@@ -94,7 +89,13 @@ if (ENABLE_LIBCXX)
         message(FATAL_ERROR "libc++ or libc++abi not found")
     endif()
 else()
-    message(STATUS "Using libstdc++")
+    if (ENABLE_CXX11_ABI)
+        add_definitions (-D_GLIBCXX_USE_CXX11_ABI=1)
+        message(STATUS "Using libstdc++ with std::__cxx11::basic_string")
```
Comment (ZhitaoShen) [inline CMakeLists.txt]: message(STATUS "Using libstdc++ with C++11 ABI")

File: CMakeLists.txt [hunk 3]
```diff
@@ -94,7 +89,13 @@ if (ENABLE_LIBCXX)
...
+    else ()
+        add_definitions (-D_GLIBCXX_USE_CXX11_ABI=0)
+        message(STATUS "Using libstdc++ with std::basic_string")
```
Comment (ZhitaoShen) [inline CMakeLists.txt]: message(STATUS "Using libstdc++ with pre-C++11 ABI")

File: Makefile [hunk 4]
```diff
@@ -27,14 +27,19 @@ debug:                  ## Build vsag with debug options.
...
 	cmake --build build --parallel ${COMPILE_JOBS}
 
 .PHONY: distribution
 distribution:           ## Build vsag with distribution options.
 	cmake ${VSAG_CMAKE_ARGS} -DCMAKE_BUILD_TYPE=Release -DENABLE_CXX11_ABI=off -DENABLE_LIBCXX=off
 	cmake --build build --parallel ${COMPILE_JOBS}
 
+.PHONY: abi11
+abi11:           ## Build vsag with abi11 options.
+	cmake ${VSAG_CMAKE_ARGS} -DCMAKE_BUILD_TYPE=Release -DENABLE_CXX11_ABI=on -DENABLE_LIBCXX=off
```
Comment (ZhitaoShen) [inline Makefile]: This means gcc with c++11 abi enabled? or this support both clang and gcc?

Comment (inabao) [reply Makefile]: Yes

File: CMakeLists.txt (hunk 1 above)
Comment (inabao) [reply CMakeLists.txt]: done

File: CMakeLists.txt (hunk 2 above)
Comment (inabao) [reply CMakeLists.txt]: done

File: CMakeLists.txt (hunk 3 above)
Comment (inabao) [reply CMakeLists.txt]: done

**Comment by ZhitaoShen** [comment]
image

Please take a look of the issue in CI.

File: CMakeLists.txt [hunk 5]
```diff
@@ -63,18 +66,14 @@ vsag_add_shared_linker_flag(-static-libstdc++)
...
-        set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -gdwarf-4")
-        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -gdwarf-4 -Wno-deprecated-builtins")
-    else()
-        set(ENABLE_LIBCXX OFF)
-    endif()
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -gdwarf-4")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -gdwarf-4 -Wno-deprecated-builtins")
 else()
-    set(ENABLE_LIBCXX OFF)
+    message(FATAL_ERROR "gcc does not support using libc++")
```
Comment (ZhitaoShen) [inline CMakeLists.txt]: If gcc then throw error?

Comment (inabao) [reply CMakeLists.txt]: I have fixed it.

**Comment by ZhitaoShen** [comment]
LGTM

**Review by wxyucs — APPROVED** [review]
lgtm

**Review by jiaweizone — APPROVED** [review]
LGTM

**Review by ZhitaoShen — APPROVED** [review]
LGTM

**Pull request merged**